readme.md
# Number Generator

Number Generator is a Jakarta-EE application that generates pseudo-random numbers and exposes a REST API.

## Pre-requisites

Before trying to run the application, ensure that you have met the following requirements:

### Java 8

Open your terminal and execute the following command to verify your Java version.

```
#java -version
openjdk version "1.8.0_232"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_232-b09)
OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.232-b09, mixed mode)
```

### Apache Maven

Type the following command to verify if Apache Maven is working.
```
#mvn -v
Maven home: /Users/inerciobelton/Downloads/apache-maven-3.6.3
Java version: 1.8.0_102, vendor: Oracle Corporation, runtime: /Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "10.12.6", arch: "x86_64", family: "mac"
```

### Jakarta EE 8 Application Server

Access the link [Jakarta EE Compatible Products](https://jakarta.ee/compatibility/) to see a full list of application servers that are compatible with the Jakarta EE 8 specification, including:

* [Wildfly 19.1.0](https://wildfly.org/downloads/)
* [Glassfish v5.1.0](https://projects.eclipse.org/projects/ee4j.glassfish/downloads)
* [OpenLiberty 19.0.0.9](https://openliberty.io/downloads/)


## Installation

To install the application locally, follow these steps:

Check out the source code to your computer:
```
git clone https://github.com/InercioBelton/number-generator.git
```
Navigate to the project directory and run the following command to build a clean package locally:
```
mvn clean install -P wildfly-managed

```
Note: The maven profile -Pwildfly-managed will start a new WildFly instance, and execute the tests, shutting it down when done. Make sure your application server is not running at this time to avoid confliting ports or adresses.
If you want to skip the tests execute the following command:
```
mvn clean install -P wildfly-managed -DskipTests

```

When the build is finished deploy the generated artifact to your favourite application server or use the DockerFile (at project root directory) to package the application in a Docker image.


## Usage

This application exposes a Rest API than can be accessed through the following endpoints:
 
* `POST /random`: [requests a new random number generation](/docs/random.md)
* `GET /history`: [get list of generated random numbers](/docs/history.md)
* `PUT /{requestId}/cancel`: [cancels a random number request](/docs/cancel.md)
* `GET stats`: [get system usage statistics](/docs/stats.md)
* `GET /pending`: [get the list of pending requests](/docs/pending.md)
* `PUT /threads`: [changes the size of the thread-pool](/docs/threads.md)
* `PUT /minRequestDuration`: [changes the minumum request duration](/docs/requestDuration.md)





docs/history.md
# Get list of generated random numbers

Returns a List of Generated Random Numbers. This includes the time that was taken to process
each random number.

**URL** : `/history`

**Example**: `http://127.0.0.1:8080/number-generator/api/history`

**Method** : `GET`

## Success Response

**Code** : `200 OK`

 

**Example**

````json

[
    {
        "generatedNumber": 334556547,
        "processingTime": "0 min, 33 sec, 33000 millis"
    },
    {
        "generatedNumber": 34232423423423,
        "processingTime": "0 min, 31 sec, 31000 millis"
    },
    {
        "generatedNumber": 5464567675668,
        "processingTime": "0 min, 42 sec, 42000 millis"
    },
    {
        "generatedNumber": 804162030458415521,
        "processingTime": "0 min, 30 sec, 30005 millis"
    }
]

````



src/main/java/com/inercio/rest/NumberGeneratorRest.java
package com.inercio.rest;

import com.inercio.model.GenericResponse;
import com.inercio.model.HistoryResponse;
import com.inercio.model.PendingResponse;
import com.inercio.model.StatsResponse;
import com.inercio.service.NumberGeneratorService;
import com.inercio.util.DelayMe;

import javax.enterprise.context.RequestScoped;
import javax.inject.Inject;
import javax.validation.constraints.Max;
import javax.validation.constraints.Min;
import javax.validation.constraints.NotNull;
import javax.ws.rs.*;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.HttpHeaders;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.util.List;

import static com.inercio.util.Constants.X_MAX_WAIT;
import static com.inercio.util.Constants.X_REQUEST_DURATION;

@RequestScoped
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
@Path("/")
public class NumberGeneratorRest {

    @Inject
    private NumberGeneratorService numberGeneratorService;

    @Context
    private HttpHeaders httpHeaders;

    @POST
    @Path("random")
    public Response generateRandomNumber() {
        long startTime = System.currentTimeMillis();
        String xMaxWait = httpHeaders.getHeaderString(X_MAX_WAIT);
        numberGeneratorService.requestRandomNumber(xMaxWait);
        GenericResponse genericResponse = new GenericResponse(200, "SUCCESS", "The random number request was submitted sucessfully");
        long elapsedTime = System.currentTimeMillis() - startTime;
        String formatedTime = numberGeneratorService.formatTime(elapsedTime);

        return Response.ok(genericResponse).header(X_REQUEST_DURATION, formatedTime).build();
    }

    @GET
    @Path("history")
    @DelayMe(time = 2000)
    public Response history() {
        long startTime = System.currentTimeMillis();
        List<HistoryResponse> randomNumberList = numberGeneratorService.getHistory();
        long elapsedTime = System.currentTimeMillis() - startTime;
        String formatedTime = numberGeneratorService.formatTime(elapsedTime);

        return Response.ok(randomNumberList).header(X_REQUEST_DURATION, formatedTime).build();
    }

    @PUT
    @Path("{requestId}/cancel")
    public Response cancel(@PathParam("requestId") String requestId) {
        long startTime = System.currentTimeMillis();
        GenericResponse genericResponse = numberGeneratorService.cancelRequest(requestId);
        long elapsedTime = System.currentTimeMillis() - startTime;

        String formatedTime = numberGeneratorService.formatTime(elapsedTime);

        return Response.ok(genericResponse).header(X_REQUEST_DURATION, formatedTime).build();

    }

    @GET
    @Path("stats")
    @DelayMe
    public Response stats() {
        long startTime = System.currentTimeMillis();
        StatsResponse statsResponse = numberGeneratorService.getStats();
        long elapsedTime = System.currentTimeMillis() - startTime;

        String formatedTime = numberGeneratorService.formatTime(elapsedTime);

        return Response.ok(statsResponse).header(X_REQUEST_DURATION, formatedTime).build();
    }

    @GET
    @Path("pending")
    public Response pending() {
        long startTime = System.currentTimeMillis();
        List<PendingResponse> pendingResponseList = numberGeneratorService.getPending();
        long elapsedTime = System.currentTimeMillis() - startTime;

        String formatedTime = numberGeneratorService.formatTime(elapsedTime);

        return Response.ok(pendingResponseList).header(X_REQUEST_DURATION, formatedTime).build();
    }

    @PUT
    @Path("threads")
    public Response threads(@QueryParam("size")
                            @Min(value = 1, message = "Minimum allowed thread pool size is 1")
                            @Max(value = 10, message = "Maximum allowed thread pool size is 10")
                            @NotNull(message = "Size cannot be null") int size) {
        long startTime = System.currentTimeMillis();
        GenericResponse response = numberGeneratorService.setThreadSize(size);
        long elapsedTime = System.currentTimeMillis() - startTime;

        String formatedTime = numberGeneratorService.formatTime(elapsedTime);
        return Response.ok(response).header(X_REQUEST_DURATION, formatedTime).build();
    }

    @PUT
    @Path("requestDuration")
    public Response requestDuration(@QueryParam("time")
                            @Min(value = 30,
                                 message = "Minimum allowed random number generation " +
                                         "request duration cannot be less than 30 seconds")
                             @NotNull(message = "Time cannot be null") long time) {
        long startTime = System.currentTimeMillis();
        GenericResponse response = numberGeneratorService.setMinRequestDuration(time);
        long elapsedTime = System.currentTimeMillis() - startTime;

        String formatedTime = numberGeneratorService.formatTime(elapsedTime);
        return Response.ok(response).header(X_REQUEST_DURATION, formatedTime).build();
    }

}
